File: cai/107.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        left=self.levelOrderBottom(root.left)
        right=self.levelOrderBottom(root.right)
        l=len(left)
        r=len(right)
        if l==0:
            return right+[[root.val]]
        if r==0:
            return left+[[root.val]]
        if r>l:
            return right[:r-l]+[left[i]+right[r-l+i] for i in range(l)]+[[root.val]]
        elif l>r:
            return left[:l-r]+[left[l-r+i]+right[i] for i in range(r)]+[[root.val]]
        else:
            return [left[i]+right[i] for i in range(l)]+[[root.val]]

File: xiong/107.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <algorithm>
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if(root == NULL) return vector<vector<int>>{};
        vector<vector<int>> ans = {};

        vector<TreeNode*> s;
        vector<int> sl;
        s.push_back(root);
        sl.push_back(0);
        while(!s.empty()){
            TreeNode* now = s.back();
            int nowl = sl.back();
            s.pop_back();
            sl.pop_back();
            
            if(nowl >= ans.size()){
                vector<int> tmp;
                tmp.push_back(now->val);
                ans.insert(ans.begin() + nowl, tmp);
            }else{
                ans[nowl].push_back(now->val);
            }
            if(now->right != NULL){
                s.push_back(now->right);
                sl.push_back(nowl + 1);
            }
            if(now->left != NULL){
                s.push_back(now->left);
                sl.push_back(nowl + 1);
            }

        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};